### PR TITLE
[RLlib] A few updates for offline DQN usage

### PR DESCRIPTION
**Pull request by gjoliver**
## Why are these changes needed?

1. Allow users to select between huber or mse losses for num_atoms=1 case.
2. Allow users to configure temperature parameter for Categorical distribution.
3. Introduce FifoReplayBuffer to allow off-policy algorithms to be used without a real ReplayBuffer.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [*] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/dqn/dqn.py [hunk 1]
```diff
@@ -246,6 +250,11 @@ def training(
                 prioritized_replay_eps: Epsilon parameter sets the baseline probability
                 for sampling so that when the temporal-difference error of a sample is
                 zero, there is still a chance of drawing the sample.
+            td_error_loss_fn: "huber" or "mse". loss function for calculating TD error
+                when num_atoms is 1. Note that if num_atoms is > 1, this parameter
+                is simply ignored, and softmax cross entropy loss will be used.
+            categorical_distribution_temperature: Set the temperature parameter used
```
Comment (kouroshHakha) [inline rllib/algorithms/dqn/dqn.py:256]: Make sure it's clear to the user that this is only during evaluation. during training we simply do argmax, right?

**Review by kouroshHakha — APPROVED** [review]
Dude, this is very nice. Thanks a ton man! just 1 nit.

File: rllib/algorithms/dqn/dqn.py (hunk 1 above)
Comment (gjoliver) [reply rllib/algorithms/dqn/dqn.py:256]: done

File: rllib/utils/replay_buffers/fifo_replay_buffer.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.replay_buffers.replay_buffer import ReplayBuffer, StorageUnit
+from ray.rllib.utils.typing import SampleBatchType
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class FifoReplayBuffer(ReplayBuffer):
+    """This replay buffer implements an FIFO queue.
```
Comment (ArturNiederfahrenhorst) [inline rllib/utils/replay_buffers/fifo_replay_buffer.py:13]: ... a FIFO queue.

File: rllib/algorithms/apex_dqn/tests/test_apex_dqn.py [hunk 3]
```diff
@@ -147,7 +147,7 @@ def _step_n_times(algo, n: int):
 
             lr = _step_n_times(algo, 3)  # 50 timesteps
             # Close to 0.2
-            self.assertGreaterEqual(lr, 0.1)
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/apex_dqn/tests/test_apex_dqn.py:150]: Finally! 😄

Comment (gjoliver) [reply rllib/algorithms/apex_dqn/tests/test_apex_dqn.py:150]: 😅

File: rllib/utils/replay_buffers/fifo_replay_buffer.py (hunk 2 above)
Comment (gjoliver) [reply rllib/utils/replay_buffers/fifo_replay_buffer.py]: done 😓

**Review by avnishn — COMMENTED** [review]
lgtm pending nit

File: rllib/utils/tf_utils.py [hunk 4]
```diff
@@ -297,6 +297,21 @@ def huber_loss(x: TensorType, delta: float = 1.0) -> TensorType:
     )
 
 
+@PublicAPI
+def l2_loss(x: TensorType) -> TensorType:
```
Comment (ArturNiederfahrenhorst) [inline rllib/utils/tf_utils.py:301]: Tensorflow also [has this loss](https://www.tensorflow.org/api_docs/python/tf/nn/l2_loss)! If there is no particular reason here for us to not use it, I think we should.

File: rllib/utils/tests/test_utils.py [hunk 5]
```diff
@@ -10,10 +10,12 @@
 from ray.rllib.utils.test_utils import check
 from ray.rllib.utils.tf_utils import (
     flatten_inputs_to_1d_tensor as flatten_tf,
+    l2_loss as tf_l2_loss,
     one_hot as one_hot_tf,
 )
 from ray.rllib.utils.torch_utils import (
     flatten_inputs_to_1d_tensor as flatten_torch,
+    l2_loss as torch_l2_loss,
```
Comment (ArturNiederfahrenhorst) [inline rllib/utils/tests/test_utils.py:18]: Just saw that we implement our own torch l2 loss, which has an mse loss https://pytorch.org/docs/master/generated/torch.nn.MSELoss.html?highlight=mse%20loss#torch.nn.MSELoss that accepts a "sum" keyword.

File: rllib/algorithms/dqn/dqn_tf_policy.py [hunk 6]
```diff
@@ -305,6 +325,8 @@ def build_q_losses(policy: Policy, model, _, train_batch: SampleBatch) -> Tensor
             q_dist_tp1 * tf.expand_dims(q_tp1_best_one_hot_selection, -1), 1
         )
 
+    loss_fn = huber_loss if policy.config["td_error_loss_fn"] == "huber" else l2_loss
```
[the remark this answers is not in the document]
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/dqn/dqn_tf_policy.py:328]: It's verified in the config object.

File: rllib/utils/replay_buffers/fifo_replay_buffer.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        Args:
+            ``*args``   : Forward compatibility args.
+            ``**kwargs``: Forward compatibility kwargs.
+
+        Returns:
+            A single training batch from the queue.
+        """
+        if len(self._queue) <= 0:
+            # Return empty SampleBatch if queue is empty.
+            return SampleBatch()
```
Comment (ArturNiederfahrenhorst) [inline rllib/utils/replay_buffers/fifo_replay_buffer.py:57]: Have you tested this? Because most of our algorithms accept MultiAgentSampleBatches!

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
Thanks Jun! Comments are all nits. _returning a SampleBatch in the buffer_ is the only thing I'd like to see resolved if it indeed presents an issue 👍

File: rllib/utils/tf_utils.py (hunk 4 above)
Comment (gjoliver) [reply rllib/utils/tf_utils.py:301]: I looked at it. tensorflow one you need to differentiate between tf1 and tf2 it seems.
I figured that's why we want to roll our own.

File: rllib/utils/tests/test_utils.py (hunk 5 above)
Comment (gjoliver) [reply rllib/utils/tests/test_utils.py:18]: what do you mean?

Comment (ArturNiederfahrenhorst) [reply rllib/utils/tests/test_utils.py:18]: Similarely to tf, we do our own math here! Not a real issue as long as the math is correct. But we had the case in another PR that we decided to better rely on Tensorbackends to do the match if they do it well!

File: rllib/utils/replay_buffers/fifo_replay_buffer.py (hunk 7 above)
Comment (gjoliver) [reply rllib/utils/replay_buffers/fifo_replay_buffer.py]: switched.
only tested with DQN tbh.
DQN uses sample_min_n_steps_from_buffer which only cares about the length of the batch.
I notice PrioritizedReplayBuffer actually raises error if the buffer is empty. I feel like that's not the best behavior here.
maybe we can discuss / revisit and make all these RBs behave similarly.

File: rllib/utils/tests/test_utils.py (hunk 5 above)
Comment (gjoliver) [reply rllib/utils/tests/test_utils.py:18]: oh I see. yeah, I don't know why we rolled these loss functions ourselves. some historical context that I don't have I think.

File: rllib/utils/replay_buffers/fifo_replay_buffer.py (hunk 7 above)
Comment (ArturNiederfahrenhorst) [reply rllib/utils/replay_buffers/fifo_replay_buffer.py]: Absolutely open to that, thanks for looking into it!

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Thanks Jun! :)
Approved if tests are green
